src/writer.rs

```diff
@@ -22,15 +23,18 @@ impl<F: LurkField> Write<F> for Ptr<F> {
             write!(w, "<Opaque ")?;
             write!(w, "{:?}", self.tag)?;
 
-            if let Some(x) = store.get_expr_hash(self) {
+            if let Some(x) = store.hash_expr(self) {
                 write!(w, " ")?;
                 crate::expr::Expression::Num(crate::num::Num::Scalar(*x.value())).fmt(store, w)?;
             }
             write!(w, ">")
         } else if let Some(expr) = store.fetch(self) {
             expr.fmt(store, w)
         } else {
-            Ok(())
+            Err(io::Error::new(
+                io::ErrorKind::Other,
+                "Cannot find expression pointer",
```

Comment:
I think that means this pointer is opaque by default. I understand that means something unexpected happened — however, I'm not convinced that a pointer having 'become opaque' should fail in this way. Something to think about.

Reply:
I think this particular error will only happen on a non-opaque pointer (there's a check at the top of the if) which `fetch` fails on.

This implies that either something is implemented wrong, or the store has been corrupted, so I think it's right to error here. Currently we assume the store is always append only, and if we want to ever support deleting things (so that pointers "become opaque) we're going to have to do a lot of work to make that safe.

Reply:
I think we might well want to do that, but I think it's fair to leave this as a problem to be solved at that time. I think it's 'reasonable' because it's within the set of things that a user of a store (as a powerful tool for data management) might want to do for some reason. But I agree we don't need to prioritize that possibility.

Reply:
The most obvious, plausible use case I can come up with off the top of my head would be that a user decides they want to especially safeguard some particular sub-graph of their data. So they explicitly remove it from the store, making it impossible that they either leak it *or* even make proofs over it. In a world where a long-running store makes sense (which should be part of some eventual use cases), then this shouldn't break their world.

Reply:
Yes, I agree that we do want to support it. It will require a substantial store refactor though, particularly because to do that safeguarding correctly you would also have to flush your caches, and it may have performance impacts wrt to our use of IndexSet.

Ultimately though I think having the store evolve in this direction is a good thing though. One option we might explore in future is doing direct allocation of store objects (similar to what @gabriel-barrett and I did in https://github.com/lurk-lab/yatima-lang-alpha/blob/main/core/src/dag.rs), keeping a hashtable to avoid duplicates, and doing actual garbage collection when we want to drop things from the store.